commit cf4fce6358f3752ad7ebd85da45c311c58cc8e65 (refs/changes/76/176/3)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2016-08-02T13:10:06-07:00 (3 years, 2 months ago)
    
    DOCKER-900: docker build: ensure chroot-gtar is only run once per directory

diff --git a/lib/build.js b/lib/build.js
index 8b9b6a3..a2ce9d7 100644
--- a/lib/build.js
+++ b/lib/build.js
@@ -27,6 +27,7 @@ var url = require('url');
 var assert = require('assert-plus');
 var async = require('async');
 var bunyan = require('bunyan');
+var chownr = require('chownr');
 var dockerFileParser = require('docker-file-parser');
 var jsprim = require('jsprim');
 var lazyProperty = require('lazy-property');
@@ -1995,40 +1996,38 @@ Builder.prototype.doCopy = function doCopy(ci, callback)
     builder.doCopyOneFile(ci, callback);
 };
 
-Builder.prototype.doCopyOneFile = function doCopyOneFile(ci, callback)
+Builder.prototype._doCopy = function _doCopy(ci, callback)
 {
     var builder = this;
-    var parentDir = path.dirname(ci.zoneDestPath);
+    var parentDir = ci.zoneDestPath;
+    if (!ci.contextPathIsDirectory) {
+        parentDir = path.dirname(ci.zoneDestPath);
+    }
 
-    builder.log.debug('copying file %j to %j', ci.contextPath, ci.zoneDestPath);
-    // Ensure parent directory exists, then copy file to it.
-    builder.mkdirpChown(parentDir, function (err) {
+    function onExtractedCb(err, result) {
         if (err) {
             callback(err);
             return;
         }
+        chownr(ci.zoneDestPath, builder.chownUid, builder.chownGid,
+            callback);
+    }
 
-        //var opts = {
-        //    'uid': builder.chownUid,
-        //    'gid': builder.chownGid,
-        //    'log': builder.log
-        //};
-        //utils.fileCopy(ci.contextPath, ci.zoneDestPath, opts, callback);
-        //return;
-
-        function onExtractedCb(err2, result) {
-            if (err2) {
-                callback(err2);
-                return;
-            }
-            fs.chown(ci.zoneDestPath, builder.chownUid, builder.chownGid,
-                callback);
+    // Ensure parent directory exists, then copy the contents to it.
+    builder.mkdirpChown(parentDir, function (err) {
+        if (err) {
+            callback(err);
+            return;
         }
-
         // Extract   /foo/bar/baz.txt    /dir/boo.txt
         // Remove leading slash to have a relative path (from root dir).
         var containerRelativePath = ci.containerAbsPath.slice(1);
         var stripDirCount = containerRelativePath.split('/').length - 1;
+        // Per docker rules, for a directory, only the contents of the directory
+        // are copied, so we need an additional strip here.
+        if (ci.contextPathIsDirectory) {
+            stripDirCount += 1;
+        }
 
         var event = {
             callback: onExtractedCb,
@@ -2043,7 +2042,8 @@ Builder.prototype.doCopyOneFile = function doCopyOneFile(ci, callback)
 
         var srcBasename = path.basename(containerRelativePath);
         var destBasename = path.basename(ci.zoneDestPath);
-        if (srcBasename !== destBasename) {
+        // Transform filename if needed (does not apply for directories).
+        if (!ci.contextPathIsDirectory && (srcBasename !== destBasename)) {
             // Regex escape the name and make a pattern like:
             // '/filename$/newname/'
             event.replacePattern = '/' + utils.escapeRegExp(srcBasename) + '$/'
@@ -2056,19 +2056,23 @@ Builder.prototype.doCopyOneFile = function doCopyOneFile(ci, callback)
     });
 };
 
+Builder.prototype.doCopyOneFile = function doCopyOneFile(ci, callback)
+{
+    var builder = this;
+
+    builder.log.debug('copying file %j to %j', ci.contextPath, ci.zoneDestPath);
+
+    builder._doCopy(ci, callback);
+};
+
 Builder.prototype.doCopyOneDirectory = function doCopyOneDirectory(ci, callback)
 {
     var builder = this;
-    builder.mkdirpChown(ci.zoneDestPath, function (err) {
-        if (err) {
-            callback(err);
-            return;
-        }
-        builder.log.debug('copying directory %j to %j',
-            ci.contextPath, ci.zoneDestPath);
-        async.eachSeries(ci.getAllChildren(), builder.doCopy.bind(builder),
-            callback);
-    });
+
+    builder.log.debug('copying directory %j to %j',
+        ci.contextPath, ci.zoneDestPath);
+
+    builder._doCopy(ci, callback);
 };
 
 Builder.prototype.reprovisionImage =
diff --git a/package.json b/package.json
index 5bd984b..c244cac 100644
--- a/package.json
+++ b/package.json
@@ -33,6 +33,7 @@
     "assert-plus": "0.1.5",
     "async": "0.9.0",
     "bunyan": "1.3.4",
+    "chownr": "1.0.1",
     "docker-file-parser": "git+https://github.com/joyent/node-docker-file-parser.git#ba6b4ce0e09a1432b200a55851d2899109d68ae2",
     "jsprim": "1.2.2",
     "lazy-property": "1.0.0",
diff --git a/test/files/addLotsOfFiles.tar.gz b/test/files/addLotsOfFiles.tar.gz
new file mode 100644
index 0000000..3caebea
Binary files /dev/null and b/test/files/addLotsOfFiles.tar.gz differ
diff --git a/test/test_build.js b/test/test_build.js
index dab3fb4..1090f42 100644
--- a/test/test_build.js
+++ b/test/test_build.js
@@ -1232,5 +1232,42 @@ tape('onbuild', function (t) {
 });
 
 
+// Ensure that adding a lot of small files doesn't error out, or take a long
+// time to complete.
+tape('addLotsOfFiles', { timeout: 60 * 1000 }, function (t) {
+    var contextFilepath = path.join(testContextDir, t.name + '.tar.gz');
+
+    testBuildContext(t, contextFilepath, function (err, result) {
+        var builder = result.builder;
+        if (showError(t, err, builder)) {
+            return;
+        }
+
+        // Verify directory contents.
+        var contents;
+        var expectedContents = [];
+        var i;
+        var root = builder.containerRootDir;
+        for (i = 1; i <= 100; i++) {
+            expectedContents.push(String(i));
+        }
+        expectedContents = expectedContents.sort();
+        for (i = 1; i <= 100; i++) {
+            contents = fs.readdirSync(path.join(root, String(i)));
+            contents = contents.sort();
+            t.deepEqual(contents, expectedContents, 'Checking dir entries');
+        }
+        // Verify the main dir (it should have an extra 'Dockerfile' filename).
+        expectedContents.push('Dockerfile');
+        expectedContents = expectedContents.sort();
+        contents = fs.readdirSync(root);
+        contents = contents.sort();
+        t.deepEqual(contents, expectedContents, 'Checking root dir entries');
+
+        testEnd(t, builder);
+    });
+});
+
+
 // Other:
 // 1. Test different command formats, i.e. 'RUN foo' and 'RUN ["foo"]'
